### ISSUE TITLE
Simplify env variable registration and deregistration

### ISSUE THREAD

**Issue by dotsdl** [xonsh/xonsh#3227]
Currently, registering an environment variable requires at least three different actions that are not exposed to the user. We want to give a single point of entry for registering and deregistering environment variables.

**Comment by NorthIsUp** [comment]
#3207

**Comment by con-f-use** [comment]
Kind of related. In bash it is possible to set an environment variable for a single command:
```
$ unset TEST
$ TEST=bla echo "$TEST" blubb
bla blubb
$ echo "$TEST" blubb
blubb
```
I know of no way to do that easily in xonsh.

**Comment by scopatz** [comment]
@con-f-use What version of Bash are you using, because that does not work in my version of Bash:

```
$ bash 
scopatz@artemis:~$ TEST=bla echo "$TEST" bluub
 bluub
scopatz@artemis:~$ bash --version
GNU bash, version 5.0.3(1)-release (x86_64-pc-linux-gnu)
```

In any event, if you need to set an environment variable at the command line in this way in xonsh, for the moment you can use the `env` command:

```
$ env TEST=bla python -c! import os; print(os.environ["TEST"])
bla
```

Note that you can't use `$NAME` because xonsh doesn't know about the env var you are setting.

**Comment by scopatz** [comment]
I think that there is an open issue for this

**Comment by con-f-use** [comment]
I just realized, it can't work, because echo is a bash built-in.  My example would work with everything but shell built-ins, e.g. python:

```
┌─22-22:36[confus@confusion6:~] $ 
└▪ echo $BASH_VERSION
4.4.20(1)-release
┌─22-22:36[confus@confusion6:~] $ 
└▪ TEST=bla python -c 'import os; print(os.environ["TEST"])'
bla
```

**Comment by scopatz** [comment]
Right, in general, that is true.  There is a way to get this to work in xonsh, since we have a lot of control over how the code is parsed and executed.  I got partway down this path at one point, but ran out of time.

**Comment by scopatz** [comment]
@dotsdl - any progress on this?  Sorry to be a nudge!

**Comment by dotsdl** [comment]
No progress so far, but hacking on this this morning. Apologies @scopatz.

And no need to apologize; I appreciate the nudge/kick. :D

**Comment by dotsdl** [comment]
Current assessment for adding a new environment variable with typing, default value, doc:
1. `Env.set_ensurer` must be called to set the type for variable.
2. There is no exposed way to set the default value in `Env._defaults`.
3. There is no exposed way to set the doc for the variable in `Env._docs`.

We will first add a single entry point for:
1. Registering a new environment variable with typing, default value, doc.
2. Deregistering an environment variable.

We will call these `Env.register` and `Env.deregister`, respectively.

Thoughts? Should these functions be exposed in a different way, or should exposure extend somewhere else? With this approach, a user can register a new envvar with (probably something like) `__xonsh__.env.register('<varname>', '<defaultval>', '<doc>')`.

**Comment by dotsdl** [comment]
PR in progress at #3377

**Comment by gforsyth** [comment]
Closed via #3377
